pocs/camera/zwo.py

```diff
@@ -0,0 +1,446 @@
+import time
+import threading
+from warnings import warn
+from contextlib import suppress
+
+from astropy import units as u
+from astropy.time import Time
+
+from pocs.camera.camera import AbstractCamera
+from pocs.camera import libasi
+from pocs.utils.images import fits as fits_utils
+from pocs.utils import error
+from pocs.utils import CountdownTimer
+from pocs.utils.logger import get_root_logger
+from pocs.utils import get_quantity_value
+
+
+class Camera(AbstractCamera):
+
+    _ASIDriver = None  # Class variable to store the ASI driver interface
+    _ids = []  # Cache of camera string IDs
+    _assigned_ids = []  # Camera string IDs already in use.
+
+    def __init__(self,
+                 name='ZWO ASI Camera',
+                 set_point=None,
+                 filter_type=None,
+                 library_path=None,
+                 timeout=0.5 * u.second,
+                 gain=None,
+                 image_type=None,
+                 *args, **kwargs):
+        # ZWO ASI cameras don't have a 'port', they only have a non-deterministic integer
+        # camera_ID and, optionally, an 8 byte ID that can be written to the camera firmware
+        # by the user (using ASICap, or pocs.camera.libasi.ASIDriver.set_ID()). We will use
+        # the latter as a serial number string.
+        kwargs['port'] = None
+        kwargs['file_extension'] = 'fits'
+        kwargs['readout_time'] = kwargs.get('readout_time', 0.1)
+
+        # Maximum time to wait beyond exposure time for an exposure to complete
+        self._timeout = get_quantity_value(timeout, unit=u.second)
+
+        # Create a lock that wil be used to prevent overlapping exposures
+        self._exposure_lock = threading.Lock()
+        self._video_event = threading.Event()
+
+        if Camera._ASIDriver is None:
+            # Initialise the driver if it hasn't already been done
+            Camera._ASIDriver = libasi.ASIDriver(library_path=library_path)
+
+        logger = get_root_logger()
+        serial_number = kwargs.get('serial_number')
+        if not serial_number:
+            msg = "Must specify serial_number (string ID) for ZWO ASI Camera"
+            logger.error(msg)
+            raise ValueError(msg)
+
+        logger.debug("Looking for ZWO ASI camera with ID '{}'".format(serial_number))
+
+        if not Camera._ids:
+            # Not cached camera IDs, need to probe for connected cameras
+            n_cameras = Camera._ASIDriver.get_num_of_connected_cameras()
+            if n_cameras == 0:
+                msg = "No ZWO ASI camera devices found"
+                logger.error(msg)
+                warn(msg)
+                return
+
+            # Get the IDs
+            for camera_index in range(n_cameras):
+                # Can get IDs without opening cameras by parsing the name string
+                self._info = Camera._ASIDriver.get_camera_property(camera_index)
+                model, _, id = self._info['name'].partition('(')
```

Comment:
Could just `split` instead of `partition` and then not worry about the `_` middle value? Or can there potentially be multiple `(` in the name?

Reply:
Also, `id` is a built-in function so this variable should change to something else just to be sure.

Reply:
The string ID part of the `name` field, enclosed in `()`, is set by the user, so for a sufficiently pathological user it may contain additional `(`...  I was careful to make zero assumptions about what is in between the first `(` and final `)`.

I could use `split` with `maxsplit=1` to only split on the first `(`  but the appeal of `partition` is that it offers a slightly neater way of handling the case where the camera hasn't had the string ID set, i.e. there is no `(` in the `name` field.

With `partition` I get an empty string back for the string ID and I can just use `if not string_ID:`. If I was using `split` I'd have to either check the length of the result or, better, catch `IndexError` when trying to access the second item of the list.